egs/wsj/s5/steps/nnet3/convert_nnet2_to_nnet3.py

```diff
@@ -0,0 +1,441 @@
+        np.savetxt(f, scales, newline='')
+        f.write(' ]')
+
+    return {'<Scales>' : filename}
+
+def ParseFixedBiasComponent(component, line, line_buffer):
+    line = ConsumeToken(component, line)
+    line = ConsumeToken("<Bias>", line)
+
+    scales = np.array([ParseVector(line)])
+
+    _, filename = tempfile.mkstemp(dir=tmpdir)
+    with open(filename, 'w') as f:
+        f.write('[ ')
+        np.savetxt(f, scales, newline='')
+        f.write(' ]')
+
+    return {'<Bias>' : filename}
+
+def ParseSpliceComponent(component, line, line_buffer):
+    line = ConsumeToken(component, line)
+    line = ConsumeToken('<InputDim>', line)
+    [input_dim, _, line] = line.strip().partition(' ')
+    line = ConsumeToken('<Context>', line)
+    context = line.strip()[1:-1].split()
+
+    return {'<InputDim>':input_dim, '<Context>':context}
+
+def ParseEndOfComponent(component, line, line_buffer):
+    # Keeps reading until it hits the end tag for component
+    end_component = '</' + component[1:]
+
+    while end_component not in line:
+        line = next(line_buffer)
+
+    return
+
+def ParseAffineComponent(component, line, line_buffer):
+    assert ("<LinearParams>" in line)
+
+    pairs = dict(re.findall('(<\w+>) ([\w.]+)', line))
+
+    # read the linear params and bias and convert it to a matrix
+    weights = ParseWeights(line_buffer)
+    bias = ParseBias(next(line_buffer))
+
+    matrix = np.concatenate([weights, bias.T], axis=1)
+
+    # write matrix and return pairs with filename
+    _, filename = tempfile.mkstemp(dir=tmpdir)
+    with open(filename, 'w') as f:
+        f.write('[ ')
+        np.savetxt(f, matrix)
+        f.write(' ]')
+
+    pairs['<Matrix>'] = filename
+
+    return pairs
+
+def ParseWeights(line_buffer):
+    weights = []
+
+    while True:
+        line = next(line_buffer)
+
+        if line.strip().endswith("["):
+            continue
+        elif line.strip().endswith("]"):
+            weights.append(ParseVector(line))
+            break
+        else:
+            weights.append(ParseVector(line))
+
+    return np.array(weights)
+
+def ParseBias(line):
+    if "<BiasParams>" in line:
+        line = ConsumeToken("<BiasParams>", line)
+
+    return np.array([ParseVector(line)])
+
+def ParseVector(line):
+    vector = line.strip().strip("[]")
+    return np.array([float(x) for x in vector.split()], dtype="float32")
+
+def ParsePriors(line, line_buffer):
+    vector = ParseVector(line.partition('[')[2])
+    priors = os.path.join(tmpdir, 'priors')
+
+    with open(priors, 'w') as f:
+        f.write('[ ')
+        np.savetxt(f, vector, newline=' ')
+        f.write(' ]')
+
+    return priors
+
+def TokenToString(token):
+    '''
+    <InputDim> -> input-dim
+    '''
+    string = token[1:-1]
+    string = re.sub(r'((?<=[a-z])[A-Z]|(?<!\A)[A-Z](?=[a-z]))', r'-\1', string).lower()
+    return string
+
+def ConsumeToken(token, line):
+    '''Returns line without token'''
+    if token != line.split()[0]:
```

Comment:
IIRC these lines can be very large. So it might be better to just check for this using regexes or at least using split() method using something like a maxsplit=1 option.